File: sdk/ts/src/vector.ts
/**
 * `Vector` — the front door to the Vector SDK.
 *
 * A `Vector` is bound once to a signing key and computes its on-chain
 * identity and PDA up front. You then authorize work in terms of plain
 * Solana **instructions** (the CPIs you want executed under the PDA); the
 * SDK wires the `advance` + `passthrough` and the transaction layout for you,
 * so you never assemble those by hand.
 *
 * Three ways to authorize, all returning ready-to-broadcast {@link Artifact}s:
 *
 * - {@link Vector.authorize} — one op.
 * - {@link Vector.chain} — an ordered, forward-secure sequence (each op can
 *   only execute after the previous one).
 * - {@link Vector.branch} — mutually-exclusive alternatives ("sign N, execute
 *   one"); whichever lands first orphans the rest.
 *
 * For independent, non-exclusive parallel work (e.g. many simultaneous RFQ
 * positions) derive sub-accounts with {@link Vector.derive} — each is another
 * `Vector` with the same API.
 *
 * Signing is **synchronous and offline** (air-gapped friendly): you pass the
 * nonce you signed against. The only networked call is {@link Vector.nonce},
 * which reads the current on-chain nonce.
 *
 * @example
 * ```ts
 * const v = Vector.ed25519(key, { feePayer });
 * const nonce = await v.nonce(connection);
 * const art = v.authorize(nonce, withdrawIx);   // op = Instruction | Instruction[]
 * await sendAndConfirmTransaction(connection, art.transaction(), [feePayer]);
 * ```
 */
import {
  Address,
  Connection,
  Transaction,
  TransactionInstruction,
} from "@solana/web3.js";

import { Scheme, findVectorPda, fetchVectorAccount } from "./scheme.js";
import {
  ED25519,
  ed25519Identity,
  createInitializeEd25519,
} from "./schemes/ed25519.js";
import { createPassthroughInstruction } from "./instructions.js";
import {
  ChainSigner,
  ed25519ChainSigner,
  signChain,
  signBranches,
  resolveChainStatus,
  deriveLaneSeed,
  BranchStep,
  SignedStep,
  ChainStatus,
} from "./branching.js";

export type { ChainStatus } from "./branching.js";

/**
 * An op — the CPI(s) to run under the Vector PDA for one authorization.
 * A single instruction or a list; an empty list is an **inert advance**
 * (bumps the nonce with no side effects, i.e. a revocation).
 */
export type Op = TransactionInstruction | TransactionInstruction[];

/** A signed, broadcast-ready authorization with the advance/passthrough wired. */
export interface Artifact {
  /** The scheme program this artifact authorizes against. */
  programId: Address;
  /** The signing identity (Ed25519: the 32-byte public key). */
  identity: Uint8Array;
  /** Nonce this artifact is signed against; valid only while the PDA sits here. */
  nonce: Uint8Array;
  /** Nonce the chain holds after this artifact executes. */
  nextNonce: Uint8Array;
  /** Fee payer the digest was bound to, if any (needed to re-verify offline). */
  feePayer?: Address;
  /** Instructions to broadcast in order: `[advance]` or `[advance, passthrough]`. */
  instructions: TransactionInstruction[];
  /** A fresh `Transaction` of {@link instructions} (relayer adds blockhash + fee-payer sig). */
  transaction(): Transaction;
}

export class Vector {
  /** The signing scheme (program) this account uses. */
  readonly scheme: Scheme;
  /** The client identity (Ed25519: the 32-byte public key). */
  readonly identity: Uint8Array;
  /** The account's PDA — its on-chain address. */
  readonly pda: Address;

  private readonly key: Uint8Array;
  private readonly signer: ChainSigner;
  private readonly feePayer?: Address;

  private constructor(
    scheme: Scheme,
    key: Uint8Array,
    signer: ChainSigner,
    pda: Address,
    feePayer?: Address
  ) {
    this.scheme = scheme;
    this.key = key;
    this.signer = signer;
    this.identity = signer.identity;
    this.pda = pda;
    this.feePayer = feePayer;
  }

  /** Bind a `Vector` to a 32-byte Ed25519 private-key seed. */
  static ed25519(key: Uint8Array, opts?: { feePayer?: Address }): Vector {
    const signer = ed25519ChainSigner(key);
    const [pda] = findVectorPda(ED25519, signer.identity);
    return new Vector(ED25519, key, signer, pda, opts?.feePayer);
  }

  /** The one-time `initialize` instruction that creates this account on-chain. */
  initialize(payer: Address): TransactionInstruction {
    return createInitializeEd25519(payer, this.identity);
  }

  /** Read the current on-chain nonce. The only networked call. */
  async nonce(connection: Connection): Promise<Uint8Array> {
    const account = await fetchVectorAccount(connection, this.scheme, this.identity);
    return account.nonce;
  }

  /**
   * Authorize a single op at `nonce`. An empty op (`[]`) is an inert advance —
   * use it to **revoke** every artifact outstanding against `nonce`.
   */
  authorize(nonce: Uint8Array, op: Op): Artifact {
    return this.chain(nonce, [op])[0];
  }

  /**
   * Authorize an ordered, forward-secure chain of ops starting at `nonce`.
   * Op `i` is signed against the nonce op `i-1` produces, so the ops can only
   * execute in order. Broadcast each returned artifact's `transaction()`.
   */
  chain(nonce: Uint8Array, ops: Op[]): Artifact[] {
    const steps = signChain(
      this.signer,
      nonce,
      ops.map((op) => this.toStep(op)),
      this.feePayer
    );
    return steps.map((s) => this.toArtifact(s));
  }

  /**
   * Authorize mutually-exclusive alternatives from one `nonce`. All share the
   * same parent state, so executing any one orphans the rest atomically.
   * Returns an artifact per label.
   */
  branch(nonce: Uint8Array, named: Record<string, Op>): Record<string, Artifact> {
    const labels = Object.keys(named);
    const branches = signBranches(
      this.signer,
      nonce,
      labels.map((label) => ({ label, steps: [this.toStep(named[label])] })),
      this.feePayer
    );
    const out: Record<string, Artifact> = {};
    branches.forEach((b, i) => {
      out[labels[i]] = this.toArtifact(b.steps[0]);
    });
    return out;
  }

  /**
   * Derive an independent sub-account (its own chain) from this account's key.
   * Use for non-exclusive parallel work (e.g. one position per RFQ deal). The
   * returned `Vector` has the same API and a distinct identity/PDA.
   */
  derive(index: number): Vector {
    const childSeed = deriveLaneSeed(this.key, "ed25519", index);
    return Vector.ed25519(childSeed, { feePayer: this.feePayer });
  }

  /**
   * Where a pre-signed chain stands given the current on-chain nonce:
   * `pending` (and which op is next), `completed`, or `orphaned`.
   */
  status(artifacts: Artifact[], currentNonce: Uint8Array): ChainStatus {
    const steps: SignedStep[] = artifacts.map((a, index) => ({
      index,
      nonce: a.nonce,
      nextNonce: a.nextNonce,
      advanceIx: a.instructions[0],
      pre: [],
      post: a.instructions.slice(1),
    }));
    return resolveChainStatus(steps, currentNonce);
  }

  /** Wrap an op's CPIs in a passthrough (or nothing, for an inert advance). */
  private toStep(op: Op): BranchStep {
    const ixs = Array.isArray(op) ? op : [op];
    if (ixs.length === 0) return {};
    return { post: [createPassthroughInstruction(this.scheme, this.identity, ixs)] };
  }

  private toArtifact(step: SignedStep): Artifact {
    const instructions = [step.advanceIx, ...step.post];
    return {
      programId: this.scheme.programId,
      identity: this.identity,
      nonce: step.nonce,
      nextNonce: step.nextNonce,
      feePayer: this.feePayer,
      instructions,
      transaction: () => new Transaction().add(...instructions),
    };
  }
}

File: crates/core/src/vector.rs
//! The transportable [`Artifact`] and the [`Op`] an authorization runs.
use solana_address::Address;
use solana_instruction::Instruction;

use crate::instructions::{
    create_advance_instruction, create_close_subinstruction, create_passthrough_instruction,
    create_withdraw_subinstruction,
};
use crate::protocol::digest::advance_vector_digest;
use crate::protocol::pda::find_vector_pda;
use crate::scheme::{Registration, Signer, SingleTxRegister};

/// What an authorization executes: nothing (inert advance / revoke), one CPI,
/// or many — all run under the vector PDA's signer seeds.
#[derive(Clone, Debug)]
pub enum Op {
    /// No-op advance: rotate the nonce only (revoke / checkpoint).
    Inert,
    /// A single CPI instruction to run under the PDA's signer seeds.
    One(Instruction),
    /// Multiple CPI instructions to run sequentially under the PDA's signer seeds.
    Many(Vec<Instruction>),
}

impl Op {
    /// Flatten the op into a `Vec<Instruction>` (empty for `Inert`).
    pub fn into_vec(self) -> Vec<Instruction> {
        match self {
            Op::Inert => vec![],
            Op::One(i) => vec![i],
            Op::Many(v) => v,
        }
    }
}

impl From<Instruction> for Op {
    fn from(i: Instruction) -> Self {
        Op::One(i)
    }
}

impl From<Vec<Instruction>> for Op {
    fn from(v: Vec<Instruction>) -> Self {
        Op::Many(v)
    }
}

/// Front door for one signing identity. Generic over the scheme's `Signer`.
pub struct Vector<S: Signer> {
    signer: S,
    fee_payer: Option<Address>,
}

impl<S: Signer> Vector<S> {
    /// Create a `Vector` with no explicit fee payer (the runtime default will be used).
    pub fn new(signer: S) -> Self {
        Self {
            signer,
            fee_payer: None,
        }
    }
    /// Create a `Vector` and record `fee_payer` in every emitted [`Artifact`].
    pub fn with_fee_payer(signer: S, fee_payer: Address) -> Self {
        Self {
            signer,
            fee_payer: Some(fee_payer),
        }
    }
    /// Client-side identity bytes for this signer (delegated to `S::identity`).
    pub fn identity(&self) -> Vec<u8> {
        self.signer.identity()
    }
    /// Derive the on-chain PDA address for this identity.
    pub fn pda(&self) -> Address {
        find_vector_pda(&S::descriptor(), &self.identity()).0
    }
}

impl<S: Registration> Vector<S> {
    /// Account-registration txs (one group per tx). Single-tx schemes return one group.
    pub fn register(&self, payer: &Address) -> Vec<Vec<Instruction>> {
        self.signer.registration_groups(payer)
    }

    /// Authorize an op against `nonce`: sign the advance digest, lay out
    /// `[..pre, advance, passthrough?]`. `Op::Inert` ⇒ advance only (revoke).
    pub fn authorize(&self, nonce: &[u8; 32], op: Op) -> Artifact {
        let identity = self.identity();
        let pre = self.signer.advance_pre_instructions();
        let ops = op.into_vec();
        let post: Vec<Instruction> = if ops.is_empty() {
            Vec::new()
        } else {
            vec![create_passthrough_instruction(
                &S::descriptor(),
                &identity,
                &ops,
            )]
        };
        let digest = advance_vector_digest(&S::descriptor(), nonce, &identity, &pre, &post);
        let sig = self.signer.sign(&digest);
        let advance = create_advance_instruction(&S::descriptor(), &identity, &sig);
        let advance_index = pre.len();
        let mut instructions = pre;
        instructions.push(advance);
        instructions.extend(post);
        Artifact {
            program_id: S::PROGRAM_ID,
            identity,
            nonce: *nonce,
            next_nonce: digest,
            fee_payer: self.fee_payer,
            public_key: self.signer.public_key(),
            advance_index,
            instructions,
        }
    }
    /// Ordered, forward-secure: each op is signed against the previous op's
    /// `next_nonce`, so step N can't land before step N-1.
    pub fn chain(&self, nonce: &[u8; 32], ops: &[Op]) -> Vec<Artifact> {
        let mut cur = *nonce;
        let mut out = Vec::with_capacity(ops.len());
        for op in ops {
            let art = self.authorize(&cur, op.clone());
            cur = art.next_nonce;
            out.push(art);
        }
        out
    }

    /// Mutually exclusive arms: every arm is signed against the same `nonce`;
    /// landing one advances the nonce and orphans the rest atomically.
    pub fn branch(
        &self,
        nonce: &[u8; 32],
        arms: std::collections::BTreeMap<String, Op>,
    ) -> std::collections::BTreeMap<String, Artifact> {
        arms.into_iter()
            .map(|(k, op)| (k, self.authorize(nonce, op)))
            .collect()
    }

    /// Authorize a `withdraw` sub-instruction: drain `lamports` from the PDA to `to`.
    pub fn withdraw(&self, nonce: &[u8; 32], to: &Address, lamports: u64) -> Artifact {
        let ix = create_withdraw_subinstruction(&S::descriptor(), &self.identity(), to, lamports);
        self.authorize(nonce, Op::One(ix))
    }
    /// Authorize a `close` sub-instruction: drain the entire PDA balance to `to`.
    pub fn close(&self, nonce: &[u8; 32], to: &Address) -> Artifact {
        let ix = create_close_subinstruction(&S::descriptor(), &self.identity(), to);
        self.authorize(nonce, Op::One(ix))
    }
}

impl<S: SingleTxRegister> Vector<S> {
    /// Convenience for single-tx schemes: the lone `initialize` instruction.
    /// Compile-time-guaranteed single group by the `SingleTxRegister` bound.
    pub fn initialize(&self, payer: &Address) -> Instruction {
        self.signer.registration_groups(payer).remove(0).remove(0)
    }
}

impl<S: crate::scheme::Derivable> Vector<S> {
    /// Independent sub-account: same API, its own deterministic key + PDA.
    pub fn derive(&self, index: u32) -> Vector<S> {
        Vector {
            signer: self.signer.derive(index),
            fee_payer: self.fee_payer,
        }
    }
}

#[cfg(all(test, feature = "ed25519"))]
mod derive_tests {
    use super::*;
    use crate::schemes::ed25519::Ed25519;
    #[test]
    fn derived_subaccounts_are_deterministic_and_distinct() {
        let root = Vector::new(Ed25519::from_seed(&[1u8; 32]));
        assert_eq!(root.derive(0).pda(), root.derive(0).pda()); // deterministic
        assert_ne!(root.derive(0).pda(), root.derive(1).pda()); // distinct
        assert_ne!(root.derive(0).pda(), root.pda()); // child != root
    }
}

#[cfg(all(test, feature = "ed25519"))]
mod facade_tests {
    use super::*;
    use crate::scheme::SchemeMeta;
    use crate::schemes::ed25519::Ed25519;
    #[test]
    fn authorize_inert_is_advance_only() {
        let v = Vector::new(Ed25519::from_seed(&[1u8; 32]));
        let art = v.authorize(&[0u8; 32], Op::Inert);
        assert_eq!(art.program_id, Ed25519::PROGRAM_ID);
        assert_eq!(art.advance_index, 0);
        assert_eq!(art.instructions.len(), 1); // inert: advance only
        assert_ne!(art.next_nonce, [0u8; 32]);
    }
    #[test]
    fn authorize_with_op_appends_passthrough() {
        let v = Vector::new(Ed25519::from_seed(&[1u8; 32]));
        let to = Ed25519::PROGRAM_ID;
        let art = v.withdraw(&[0u8; 32], &to, 1);
        assert_eq!(art.instructions.len(), 2); // advance + passthrough
        assert_eq!(art.advance_index, 0);
    }
    #[test]
    fn register_single_group_for_curve() {
        let v = Vector::new(Ed25519::from_seed(&[1u8; 32]));
        let groups = v.register(&Ed25519::PROGRAM_ID);
        assert_eq!(groups.len(), 1);
        assert_eq!(groups[0].len(), 1);
    }
}

#[cfg(all(test, feature = "ed25519"))]
mod branch_tests {
    use super::*;
    use crate::schemes::ed25519::Ed25519;
    #[test]
    fn chain_links_nonces() {
        let v = Vector::new(Ed25519::from_seed(&[1u8; 32]));
        let arts = v.chain(&[0u8; 32], &[Op::Inert, Op::Inert]);
        assert_eq!(arts.len(), 2);
        assert_eq!(arts[1].nonce, arts[0].next_nonce); // step 2 signed against step 1's result
        assert_ne!(arts[0].nonce, arts[1].nonce);
    }
    #[test]
    fn branches_share_nonce() {
        use std::collections::BTreeMap;
        let v = Vector::new(Ed25519::from_seed(&[1u8; 32]));
        let mut arms = BTreeMap::new();
        arms.insert("a".to_string(), Op::Inert);
        arms.insert("b".to_string(), Op::Inert);
        let out = v.branch(&[7u8; 32], arms);
        assert_eq!(out["a"].nonce, out["b"].nonce); // same source nonce
        assert_eq!(out.len(), 2);
    }
}

/// A signed, transportable authorization. `instructions` is the full ordered
/// layout `[..pre, advance, ..passthrough]`; `advance_index` is where the
/// advance sits.
#[derive(Clone, Debug, PartialEq)]
pub struct Artifact {
    /// Program ID of the scheme that produced this artifact.
    pub program_id: Address,
    /// Client-side identity bytes (pubkey or `sha256(wire_pk)` for PQ schemes).
    pub identity: Vec<u8>,
    /// Nonce the signer committed to; becomes stale once the advance lands.
    pub nonce: [u8; 32],
    /// The advance digest: becomes the new on-chain nonce after the advance lands.
    pub next_nonce: [u8; 32],
    /// Optional explicit fee payer; `None` means the runtime default.
    pub fee_payer: Option<Address>,
    /// Wire public key (PQ schemes only); `None` for curve schemes.
    pub public_key: Option<Vec<u8>>,
    /// Index of the `advance` instruction within `instructions`.
    pub advance_index: usize,
    /// Full ordered instruction layout: `[..pre, advance, ..passthrough]`.
    pub instructions: Vec<Instruction>,
}
